lib/mapping/CCampaignHandler.cpp

```diff
@@ -151,6 +180,257 @@ std::string CCampaignHandler::readLocalizedString(CBinaryReader & reader, std::s
+	{
+		CCampaignScenario::SScenarioPrologEpilog ret;
+		ret.hasPrologEpilog = !identifier.isNull();
+		if(ret.hasPrologEpilog)
+		{
+			ret.prologVideo = identifier["video"].Integer();
+			ret.prologMusic = identifier["music"].Integer();
+			ret.prologText = identifier["text"].String();
+		}
+		return ret;
+	};
+
+	CCampaignScenario ret;
+	ret.conquered = false;
+	ret.mapName = reader["map"].String();
+	for(auto & g : reader["precoditions"].Vector())
+		ret.preconditionRegions.insert(g.Integer());
+
+	ret.regionColor = reader["color"].Integer();
+	ret.difficulty = reader["difficulty"].Integer();
+	ret.regionText = reader["regionText"].String();
+	ret.prolog = prologEpilogReader(reader["prolog"]);
+	ret.epilog = prologEpilogReader(reader["epilog"]);
+
+	ret.travelOptions = readScenarioTravelFromJson(reader, version);
+
+	return ret;
+}
+
+CScenarioTravel CCampaignHandler::readScenarioTravelFromJson(JsonNode & reader, int version )
+{
+	CScenarioTravel ret;
+
+	std::map<std::string, ui8> heroKeepsMap = {
+		{"experience", 1},
+		{"primarySkill", 2},
+		{"secondarySkill", 4},
+		{"spells", 8},
+		{"artifacts", 16}
+	};
+	std::map<std::string, ui8> startOptionsMap = {
+		{"none", 0},
+		{"bonus", 1},
+		{"crossover", 2},
+		{"hero", 3}
+	};
+	
+	std::map<std::string, CScenarioTravel::STravelBonus::EBonusType> bonusTypeMap = {
+		{"spell", CScenarioTravel::STravelBonus::EBonusType::SPELL},
+		{"creature", CScenarioTravel::STravelBonus::EBonusType::MONSTER},
+		{"building", CScenarioTravel::STravelBonus::EBonusType::BUILDING},
+		{"artifact", CScenarioTravel::STravelBonus::EBonusType::ARTIFACT},
+		{"scroll", CScenarioTravel::STravelBonus::EBonusType::SPELL_SCROLL},
+		{"primarySkill", CScenarioTravel::STravelBonus::EBonusType::PRIMARY_SKILL},
+		{"secondarySkill", CScenarioTravel::STravelBonus::EBonusType::SECONDARY_SKILL},
+		{"resource", CScenarioTravel::STravelBonus::EBonusType::RESOURCE},
+		//{"prevHero", CScenarioTravel::STravelBonus::EBonusType::HEROES_FROM_PREVIOUS_SCENARIO},
+		//{"hero", CScenarioTravel::STravelBonus::EBonusType::HERO},
+	};
+	
+	std::map<std::string, ui32> primarySkillsMap = {
+		{"attack", 0},
+		{"defence", 8},
+		{"spellpower", 16},
+		{"knowledge", 24},
+	};
+	
+	std::map<std::string, ui16> heroSpecialMap = {
+		{"strongest", 0xFFFD},
+		{"generated", 0xFFFE},
+		{"random", 0xFFFF}
+	};
+	
+	std::map<std::string, ui8> resourceTypeMap = {
+		//FD - wood+ore
+		//FE - mercury+sulfur+crystal+gem
+		{"wood", 0},
+		{"mercury", 1},
+		{"ore", 2},
+		{"sulfur", 3},
+		{"crystal", 4},
+		{"gems", 5},
+		{"gold", 6},
+		{"common", 0xFD},
+		{"rare", 0xFE}
+	};
+	
+	for(auto & k : reader["heroKeeps"].Vector())
+		ret.whatHeroKeeps |= heroKeepsMap[k.String()];
+	
+	for(auto & k : reader["keepCreatures"].Vector())
+	{
+		if(auto identifier = VLC->modh->identifiers.getIdentifier(CModHandler::scopeMap(), "creature", k.String()))
+		{
+			int creId = identifier.get();
+			ret.monstersKeptByHero[creId / 8] |= (1 << creId % 8);

```

Comment:
this is std::array at the moment. Meaning, it will overflow & crash for creatures from mods

(same for artifsKeptByHero below)